openmicroscopy/openmicroscopy
# Advanced email configuration

To follow-on from the recent mail improvements, this PR introduces a few build modifications as well as listener objects to allow for a more sysadmin-directed configuration of system notifications. One of the main benefits is that it is now possible to write **integration tests** for mail behaviors.

### Deployment changes ###
 * Support `classpath*:blitz/*.xml` such that Spring bean files can be added under etc/ and be picked up automatically.
 * Support `export OMERO_PROFILE=/opt/omero-config/` as another location for XML (Note the ending slash)

### Server-side code changes ###
 * Moved MailUtil to new `ome.services.mail` package (cc: @aleksandra-tarkowska)
 * Remove `ShapeChangeMessage` (unused)
 * Remove old spring xml files
 * Add `SqlAction.getUserEmailsByGroup()`
 * Add a `ShutdownMessage` raised in `ome.services.blitz.Entry`
 * Add `subethasmtp.jar` dependency, by default the runtime doesn't access it. Uncommenting or copying a blob from `spring-cfg/mail.xml` to some other location will start up a SMTP server.

### Functional email changes ###
 * Send emails during certain events **if activated**:
    * When a new user is added @ehrenfeu
    * When a server starts
    * When a server shuts down

### Test setup ###
 * Make a new directory for placing your config files (e.g. /tmp/cfg) and subdirectory blitz (/tmp/cfg/blitz/)
 * Set that directory as a location for loading config, `export OMERO_PROFILE=/tmp/cfg/`
 * Copy the two files from etc/blitz/ and rename:
   * `mv etc/blitz/mail-server.example /tmp/cfg/blitz`
   * `mv etc/blitz/mail-senders.example /tmp/cfg/blitz`
 * Restart the server
 * Make sure `root` has a password: `bin/omero obj update Experimenter:0 email=test@localhost`

### Testing ###
 * Run: `test/integration/test_mail.py`. Note: this is being run by the CI infrastructure as well.
 * Verify that one of the actions configured (comment on image or new user) create a MapAnnotation as expected.

--no-rebase
